### ISSUE TITLE
Feature request: slow sinc interpolation at unevenly spaced points

### ISSUE THREAD

**Issue by antoine-levitt** [roflmaostc/FFTResampling.jl#1]
Would be nice to have this work for arbitrary points, also multi-dimensional. I have this piece of code lying around:
```
# take an array ϕ defined at equispaced points and evaluate its
# Fourier interpolant at points xint
function fourier_interpolate(ϕ,xint)
    Y = fftshift(fft(ϕ))
    half_size = floor.(size(ϕ)./2)
    yval = zeros(size(xint))
    x = [[x[i] for x in xint] for i in 1:3]
    for k = 1:size(ϕ)[3], j = 1:size(ϕ)[2], i = 1:size(ϕ)[1]
	yval += Y[i,j,k]*exp.(im*2*pi*
			      ((i-1 - half_size[1]).*x[1] .+
			       (j-1 - half_size[2]).*x[2] .+
			       (k-1 - half_size[3]).*x[3]))
    end
    return real.(yval ./length(ϕ))
end
```

**Comment by roflmaostc** [comment]
You mean N dimensional signals and evaluating them at non-equidistant spaced points?

I believe it could be implemented in N dimensions but the computation gets pretty costly because the function evaluation takes O(M) time where M is the number of total points.
However, it might be the case that for higher dimensions the FFT (to get the coefficients) is the limiting factor?

**Comment by roflmaostc** [comment]
## Runtime evaluation
### Equdistant interpolation via zero padding in frequency domain
Let's assume we have M points in each dimension. The most expensive part is the FFT. Therefore total complexity is O( (M ^ k * log(M^k)) = O(M^k * k *log(M)).

### Sinc sum interpolation
First we need a Fourier transform to get the coefficients.
Second, we need to evaluate the sum. The sum evaluation costs for each point O(M^k) because naively we need all points of all dimensions.
Third we need this evaluation for all new points (L per dimension) O(L^k). Therefore in total to calculate all new points O(L^k * M^k).
In total for the whole routine: O(M^k * k *log(M)) + O(L^k * M^k) = O(L^k * M^k)

**Comment by antoine-levitt** [comment]
I'm confused by your reasoning. What I mean is that you've got N equispaced data points, and want to evaluate their interpolation at random points. That's what the code I posted above does. This is O(M N) in 1 dimension, it's O(data_points * evaluation_points) in any dimension. Plus some FFTs but log factors are irrelevant anyway and the FFT will not dominate.

**Comment by antoine-levitt** [comment]
Of course it's very slow, that's why it's rarely useful and evaluating on equispaced grids is much better, but sometimes you don't have a choice, and it's useful as a fallback/debug

**Comment by roflmaostc** [comment]
Updated it, sorry for the mistakes. 
If I've got time I might tackle this.

**Comment by antoine-levitt** [comment]
I agree with the new version. It's O(data_points * evaluation_points) as is clear from the code I posted. Note that it's old code - it might only work for even/odd data, be slow, etc.

**Comment by roflmaostc** [comment]
Yeah, reasoning in editor is hard :D
Currently I prioritize the correct down/upsampling in Fourier space (getting the boundary components right).
